### ISSUE TITLE
Sunshine no longer working with Wayland/headless backends

### ISSUE THREAD

**Issue by phonetic112** [hyprwm/Hyprland#6623]
### Regression?

Yes

### System Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit 4778afe2e6b4a6f8c7d218ccd8fe7e0bd4d2ee9c  (hyprctl: make recv timeout bigger and give error message if it does timeout (6621)).
Date: Fri Jun 21 18:41:23 2024
Tag: v0.41.1-56-g4778afe2, commits: 4874

flags: (if any)


System Information:
System name: Linux
Node name: arch
Release: 6.9.5-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Sun, 16 Jun 2024 19:06:37 +0000


GPU information: 
07:00.0 VGA compatible controller [0300]: NVIDIA Corporation AD104 [GeForce RTX 4070] [10de:2786] (rev a1) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

```

</details>

### Description

Sunshine can't select the monitor if the Wayland or headless backends are used. Regressed by 4f26ae70fde3b4cd5ad8ef096a7f0cb6c2dc5528.

### How to reproduce

Run sunshine in a nested session with `output_name = 1`

### Crash reports, logs, images, videos

[sunshine.log](https://github.com/user-attachments/files/15934364/sunshine.log)

**Comment by vaxerski** [comment]
`[2024:06:21:19:13:48]: Error: Unknown Monitor connector type [WL]: Please report this to the GitHub issue tracker`

Is this a problem with us? from this log I cannot tell.

`[2024:06:21:19:13:48]: Warning: Missing Wayland wire for wlr-export-dmabuf` sure is missing, you should use screencopy instead.

**Comment by phonetic112** [comment]
since wlr_export_dmabuf isn't available it's falling back to drm/kms capture, which doesn't work with headless outputs

**Comment by vaxerski** [comment]
it should use wlr_screencopy and not export_dmabuf. I have no plans on fixing this.

**Comment by martydingo** [comment]
I also want to add the reasoning for dmabuf, as found on an issue discussing NVFBC and X11

> For Nvidia, yes. For AMD and Intel, no.

> The reason is that Nvidia doesn't support VAAPI encoding, so KMS/X11/wlgrab frames need to be read back from the GPU to CPU memory for NVENC encoding. When using KMS/wlgrab+VAAPI or NvFBC+NVENC, we can directly import GPU buffers for encoding without having to read them back to the CPU first.

Which seems reasonable and disagrees with the commit statement.

**Comment by vaxerski** [comment]
wlr_export_dmabuf has been superseded years ago by wlr_screencopy_v1

**Comment by martydingo** [comment]
Could it be noted somewhere? It is a breaking change for Sunshine and there will be people in the same boat as me probably trying to understand what their issue is.

**Comment by vaxerski** [comment]
it's noted here, if they encounter the same issue they should search the bug tracker and find this issue.

**Comment by plantroon** [comment]
If you hit this from search engines (I did), look at: https://github.com/LizardByte/Sunshine/issues/2955#issuecomment-2694752564

**Comment by mirql** [comment]
hello, ive also fight with this problem and in result all great and working, thanks @plantroon 

so, im on nixos and trying to use sunshine with hyprland with virtual/headless monitor

on my laptop 2 gpus - igpu intel and discrete nvidia

so there was alot of problems ill try to show my current setup:

```nix
  
  hardware.display.edid.enable = true;
  # DP-3 is a real output that connected to the same output as my main eDP-1
  # check below what output you should use in your case
  hardware.display.outputs."DP-3".mode = "2560x1600@60e";
  
  # this edid i got from my real current display
  hardware.display.outputs."DP-3".edid = "custom1.bin";

  hardware.display.edid.packages = [
    (pkgs.runCommand "edid-custom" { } ''
      mkdir -p "$out/lib/firmware/edid"

      # this edid you can copy from your real monitor, check below

      base64 -d > "$out/lib/firmware/edid/custom1.bin" <<'EOF'
      AP///////wBRuAMWAAAAAA8gAQS1IxZ4Bw+RrlJDsCYPUFQAAAABAQEBAQEBAQEBAQEBAQEBhXAA
      oKBApWAwIDYAWdcQAAAaAAAA/Qw88KiocwEKICAgICAgAAAA/ABUTDE2MEFETVAwMy0wAAAAEAww
      8FBQQwAKICAgICAgAatwIHkCACIAFL+UEYX/CZ8ALwAfAD8GpAACAAUAKwAGJwA87wAAAAAAAAAA
      AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
      AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADekA==
      EOF
    '')

  boot.kernelParams = [
    # that is my real output eDP-1
    "video=eDP-1:2560x1600@240e"
    # this one we are activating by force with "e"
    "video=DP-3:2560x1600@240e"
  ];


```


also this can help:

**1. edid**: here we can find where is our "edid's"

we need to take 

```bash
λ sudo find /sys |grep -i edid
/sys/kernel/debug/dri/0000:01:00.0/eDP-1/edid_override
/sys/kernel/debug/dri/0000:01:00.0/HDMI-A-1/edid_override
/sys/kernel/debug/dri/0000:01:00.0/DP-3/edid_override
/sys/kernel/debug/dri/0000:00:02.0/DP-2/edid_override
/sys/kernel/debug/dri/0000:00:02.0/DP-1/edid_override
/sys/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2/card2-HDMI-A-1/edid
# this one is activated by configs so we see that used same edid as a real display
/sys/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2/card2-DP-3/edid
# so this one is real display:
/sys/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2/card2-eDP-1/edid
/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-2/edid
/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-1/edid
/sys/module/drm/parameters/edid_firmware
/sys/module/drm/parameters/edid_fixup
```

get your edid in base64 form to put in above config:

```
λ base64 < /sys/devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2/card2-eDP-1/edid
AP///////wBRuAMWAAAAAA8gAQS1IxZ4Bw+RrlJDsCYPUFQAAAABAQEBAQEBAQEBAQEBAQEBhXAA
oKBApWAwIDYAWdcQAAAaAAAA/Qw88KiocwEKICAgICAgAAAA/ABUTDE2MEFETVAwMy0wAAAAEAww
8FBQQwAKICAgICAgAatwIHkCACIAFL+UEYX/CZ8ALwAfAD8GpAACAAUAKwAGJwA87wAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADekA==
```

**2. output to use**: here we can see list of outputs that we can use

```bash
λ ls -la /sys/class/drm
lrwxrwxrwx    - root 10 Mar 00:00 card1 -> ../../devices/pci0000:00/0000:00:02.0/drm/card1
lrwxrwxrwx    - root 10 Mar 00:00 card1-DP-1 -> ../../devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-1
lrwxrwxrwx    - root 10 Mar 00:00 card1-DP-2 -> ../../devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-2
lrwxrwxrwx    - root 10 Mar 00:00 card2 -> ../../devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2
lrwxrwxrwx    - root 10 Mar 00:00 card2-DP-3 -> ../../devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2/card2-DP-3
lrwxrwxrwx    - root 10 Mar 00:00 card2-eDP-1 -> ../../devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2/card2-eDP-1
lrwxrwxrwx    - root 10 Mar 00:00 card2-HDMI-A-1 -> ../../devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/card2/card2-HDMI-A-1
lrwxrwxrwx    - root 10 Mar 00:00 renderD128 -> ../../devices/pci0000:00/0000:00:02.0/drm/renderD128
lrwxrwxrwx    - root 10 Mar 00:00 renderD129 -> ../../devices/pci0000:00/0000:00:01.0/0000:01:00.0/drm/renderD129
.r--r--r-- 4.1k root 10 Mar 00:00 version
```

i guess we need to use same output as a main display output so ive took "DP-3", same "card2"


if have any questions if i can help just let me now, my hardware is asus zephyrus m16 laptop


@martydingo

**Comment by mirql** [comment]
@vaxerski thanks for you awesome hyprland project, using it for years, best one :3

i came from awesome -> i3 world before hyprland and now im here and happier than before

**Comment by nomisma-qt** [comment]
Hi, i'm trying to get a Fedora 41 wayland desktop working as headless. 
[https://github.com/LizardByte/Sunshine/issues/2250#issuecomment-2831524527](url)

Regardless of what i do, i only get unknown monitor with resolution of 1024x768.

Is there any fix as of now? nvidia driver is 570.144. I found a link suggesting it's fixed in 575.51.02 Beta?
[https://forums.developer.nvidia.com/t/overriding-edid-makes-vrr-stop-working-under-wayland-vrr-capable-immutable-range-0-1-0/302929/33](url)

very thankful for any help

**Comment by Nosirus** [comment]
For my part, the EDID method works well, like with a fake HDMI dongle, but how can I prevent the real screen from being reactivated when connecting with Sunshine? I use Hyprland, Hypridle, and I've tried everything, but it doesn't replicate the headless principle that keeps the real display turned off.
